File: two_pointers/02_two_integer_sum/Main.java
import java.util.HashMap;
import java.util.Map;

class Main{

    public int[] custom(int[]numbers, int target){
        Map<Integer, Integer> seen = new HashMap<>();

        for (int i = 0; i < numbers.length; i++){
            int diff = target - numbers[i];
            if (seen.containsKey(diff)){
                return new int []{
                    seen.get(diff) + 1, i + 1
                };
            }
            seen.put(numbers[i], i);
        }
        return new int[]{};
    }


    public int[] bruteForce(int[] numbers, int target) {
        for (int i = 0; i < numbers.length; i++) {
            for (int j = i + 1; j < numbers.length; j++) {
                if (numbers[i] + numbers[j] == target){
                    return new int[] {i + 1, j +1};
                }
            }
        }
        return new int[0];
    }

    
    public int[] binarySearch(int[] numbers, int target) {
        for (int i = 0; i < numbers.length; i++) {
            int l = i +1, r= numbers.length - 1;
            int tmp = target - numbers[i];

            while (l <= r) {
                int mid = l + (r - l) / 2;
                if (numbers[mid] == tmp){
                    return new int[]{i + 1, mid + 1};
                }else if (numbers[mid] < tmp){
                    l = mid + 1;
                }else{
                    r = mid - 1;
                }
            }
        }
        return new int[0];
    }


    public int[] hashMap(int[] numbers, int target) {
        Map<Integer, Integer> mp = new HashMap<>();
        for (int i = 0; i < numbers.length; i++){
            int tmp = target - numbers[i];
            if (mp.containsKey(tmp)){
                return new int []{mp.get(tmp), i + 1};
            }
            mp.put(numbers[i], i + 1);
        }
        return new int[0];
    }


    public int[] twoPointers(int[] numbers, int target) {
        int l = 0, r = numbers.length - 1;

        while (l < r){
            int curSum = numbers[l] + numbers[r];

            if (curSum > target){
                r --;
            }else if (curSum < target){
                l++;
            }else{
                return new int[]{l + 1, r + 1};
            }
        }
        return new int[0];
    }

}

File: two_pointers/02_two_integer_sum/main.py
from typing import List, DefaultDict

def custom(numbers:List[int], target:int) -> List[int]:
    seen = {}
    for index, number in enumerate(numbers):
        diff = target - number
        if diff in seen:
            return [seen[index] + 1, index + 1]
        seen[number] = index 
    return [] 


def brute_force(numbers:List[int], target:int) -> List[int]:
    for i in range(len(numbers)):
        for j in range(i + 1, len(numbers)):
            if numbers[i] + numbers[j] == target:
                return [i + 1, j + 1]
    return []


def binary_search(numbers: List[int], target:int) -> List[int]:
    for i in range(len(numbers)):
        l, r = i + 1, len(numbers) - 1
        tmp = target - numbers[i]
        while l <= r:
            mid = l + (r - l ) // 2
            if numbers[mid] == tmp:
                return [i + 1, mid + 1]
            elif numbers[mid] < tmp:
                l = mid + 1

            else:
                r = mid - 1
    return []


def hash_map(numbers: List[int], target:int) -> List[int]:
    mp = DefaultDict(int)
    for i in range(len(numbers)):
        tmp = target - numbers[i]
        if mp[tmp]:
            return [mp[tmp, i+1]]
        mp[numbers[i]] = i + 1
    return []


def two_pointers(numbers: List[int], target:int) -> List[int]:
    l, r =0, len(numbers) - 1
    while l < r:
        curSum = numbers[l] + numbers[r]
        if  curSum > target:
            r -= 1
        elif curSum < target:
            l += 1
        else:
            return [l + 1, r + 1]
    return []